File: backend/app/schemas/knowledge.py
"""Knowledge Graph Visualization & Traversal Schemas."""

from __future__ import annotations

from typing import Any
from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    type: str
    label: str
    properties: dict[str, Any] = {}


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relationship: str
    properties: dict[str, Any] = {}


class GraphData(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class EntityDetail(BaseModel):
    id: str
    name: str
    type: str
    properties: dict[str, Any]
    neighbors: GraphData


class TraversalRequest(BaseModel):
    entity_id: str
    max_depth: int = 2


class TraversalResponse(BaseModel):
    graph: GraphData


class ArchitectureServiceBlock(BaseModel):
    id: str
    name: str
    tier: str
    language: str | None = "Go"
    description: str | None = "Enterprise microservice component"
    owner_team: str | None = "Platform Team"
    lead_contact: str | None = "james.liu@acmeai.com"
    apis: list[str] = []
    dependencies: list[str] = []
    downstream_services: list[str] = []
    upstream_services: list[str] = []
    related_repositories: list[str] = []
    related_incidents: list[str] = []
    documentation: list[str] = []
    environment: str = "Production"
    criticality: str = "High"
    recent_changes: list[str] = []


class ArchitectureOverviewResponse(BaseModel):
    services: list[ArchitectureServiceBlock]
    tiers: list[str]
    departments: list[str]
    total_services: int
    total_dependencies: int


class AffectedEntitySummary(BaseModel):
    id: str
    name: str
    type: str  # Service, API, Team, Repository, Incident, Document
    impact_level: str  # Direct, Downstream, Systemic
    criticality: str  # High, Medium, Low


class ImpactAnalysisResponse(BaseModel):
    target_entity: str
    target_type: str
    blast_radius_depth: int
    criticality_score: float  # e.g. 9.4 out of 10
    overall_risk_level: str  # HIGH, MEDIUM, LOW
    total_dependent_services: int
    affected_teams_count: int
    impacted_apis_count: int
    incident_history_count: int
    affected_services: list[AffectedEntitySummary]
    affected_apis: list[AffectedEntitySummary]
    affected_teams: list[AffectedEntitySummary]
    related_repositories: list[AffectedEntitySummary]
    recent_incidents: list[AffectedEntitySummary]
    critical_documents: list[AffectedEntitySummary]
    security_risks: list[str]
    impact_graph: GraphData


class ShortestPathResponse(BaseModel):
    source_id: str
    target_id: str
    path_found: bool
    path_nodes: list[GraphNode]
    path_edges: list[GraphEdge]
    total_hops: int
    description: str


class SearchResultItem(BaseModel):
    id: str
    name: str
    type: str
    snippet: str | None = None
    properties: dict[str, Any] = {}


class GraphSearchResponse(BaseModel):
    query: str
    total_results: int
    results: list[SearchResultItem]


class EdgeInspectionDetails(BaseModel):
    edge_id: str
    source_id: str
    source_name: str
    target_id: str
    target_name: str
    relationship: str
    confidence: float = 0.98
    description: str
    supporting_document: str = "acme-architecture-overview.md"
    extracted_evidence: str
    similarity_score: float = 0.96
    cypher_query: str
    date_discovered: str = "2025-06-15"
    last_verified: str = "Today"

File: frontend/src/types/knowledge.ts
export interface GraphNode {
  id: string;
  type: string;
  label: string;
  properties?: Record<string, any>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  relationship: string;
  properties?: Record<string, any>;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface ArchitectureServiceBlock {
  id: string;
  name: string;
  tier: string;
  language?: string;
  description?: string;
  owner_team?: string;
  lead_contact?: string;
  apis: string[];
  dependencies: string[];
  downstream_services: string[];
  upstream_services: string[];
  related_repositories: string[];
  related_incidents: string[];
  documentation: string[];
  environment: string;
  criticality: string;
  recent_changes: string[];
}

export interface ArchitectureOverviewResponse {
  services: ArchitectureServiceBlock[];
  tiers: string[];
  departments: string[];
  total_services: number;
  total_dependencies: number;
}

export interface AffectedEntitySummary {
  id: string;
  name: string;
  type: string;
  impact_level: string;
  criticality: string;
}

export interface ImpactAnalysisResponse {
  target_entity: string;
  target_type: string;
  blast_radius_depth: number;
  criticality_score: number;
  overall_risk_level: 'HIGH' | 'MEDIUM' | 'LOW' | string;
  total_dependent_services: number;
  affected_teams_count: number;
  impacted_apis_count: number;
  incident_history_count: number;
  affected_services: AffectedEntitySummary[];
  affected_apis: AffectedEntitySummary[];
  affected_teams: AffectedEntitySummary[];
  related_repositories: AffectedEntitySummary[];
  recent_incidents: AffectedEntitySummary[];
  critical_documents: AffectedEntitySummary[];
  security_risks: string[];
  impact_graph: GraphData;
}

export interface ShortestPathResponse {
  source_id: string;
  target_id: string;
  path_found: boolean;
  path_nodes: GraphNode[];
  path_edges: GraphEdge[];
  total_hops: number;
  description: string;
}

export interface SearchResultItem {
  id: string;
  name: string;
  type: string;
  snippet?: string;
  properties?: Record<string, any>;
}

export interface GraphSearchResponse {
  query: string;
  total_results: number;
  results: SearchResultItem[];
}

export interface EdgeInspectionDetails {
  edge_id: string;
  source_id: string;
  source_name: string;
  target_id: string;
  target_name: string;
  relationship: string;
  confidence: number;
  description: string;
  supporting_document: string;
  extracted_evidence: string;
  similarity_score: number;
  cypher_query: string;
  date_discovered: string;
  last_verified: string;
}
